Split FileManager into Platform Dependent Files

jmittert opened the pull request:
Splits the Windows and POSIX specific File Manager implementations into their own files. See https://github.com/apple/swift-corelibs-foundation/pull/2002#issuecomment-477242876

Not sure if there's a way to make the diff more readable, but in general, what I've done is copy over the function to win32/POSIX, prefixed the name with an underscore, removed the non applicable #if branch, and have the FileManager version pass through to it.

spevans commented:
Might be worth moving `internal class NSURLDirectoryEnumerator : DirectoryEnumerator` to the per OS files as well as there looks to be very little shared there.

spevans commented:
@swift-ci test

spevans commented on Foundation/FileManager+POSIX.swift:
```diff
@@ -0,0 +1,778 @@
+#endif
+        return urls
+    }
+ internal func _urls(for directory: SearchPathDirectory, in domainMask: SearchPathDomainMask) -> [URL] {
```
I think the per-OS helper functions should be `private` instead of `internal` where possible`. These are probably a holdover from when they were first written but may as well fix them now.

jmittert replied:
I believe most of these need to be internal instead of private since they are used in a file they aren't defined in.

spevans replied:
Ah yeah, good point.

spevans commented on Foundation/FileManager+POSIX.swift:
```diff
@@ -0,0 +1,778 @@
+
+        return urls
+    }
+    internal func _createDirectory(atPath path: String, withIntermediateDirectories createIntermediates: Bool, attributes: [FileAttributeKey : Any]? = [:]) throws {
```
Add a separating blank line between functions

spevans commented on Foundation/FileManager+POSIX.swift:
```diff
@@ -0,0 +1,778 @@
+             throw _NSErrorWithErrno(errno, reading: false, path: path)
+         }
+     })
+ }
```
This needs indentation.

spevans commented on Foundation/FileManager+POSIX.swift:
```diff
@@ -0,0 +1,778 @@
+            return
+        }
+
+        if rmdir(path) == 0 {
```
This seems to have lost its `_fileSystemRepresentation` wrapper from #1994. Did you base this split on the latest version from `master`?

jmittert commented:
- Did some formatting
- Fixed the missing fsr, After going through, I pretty sure that it was the only one that got dropped...
- Also split out the iterators

spevans commented:
@swift-ci test

spevans commented:
I think the following can be moved into `FileManager+POSIX.swift` as well

```
darwinPathURLs
darwinURLS
xdgHomeDirectory
xdgURLs
```

and wrapped inside a `#if canImport(Darwin) / #else / #endif` block

jmittert commented:
> I think the following can be moved into `FileManager+POSIX.swift` as well
> 
> ```
> darwinPathURLs
> darwinURLS
> xdgHomeDirectory
> xdgURLs
> ```
> and wrapped inside a `#if canImport(Darwin) / #else / #endif` block

Moved, though not in an `#if` since they're used in places not guarded by an `#if`

millenomi commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

millenomi commented:
@jmittert I'm okay taking this if you fix the conflicts with the things that are landing.

jmittert commented:
Alright, rebased!

spevans commented:
@swift-ci test

compnerd commented:
@swift-ci please test

jmittert commented:
These both failed on the Swift test suite
```
11:11:25 ********************
11:11:25 Failing Tests (2):
11:11:25     Swift(macosx-x86_64) :: IRGen/autorelease_optimized_armv7.sil
11:11:25     Swift(macosx-x86_64) :: IRGen/autorelease_optimized_aarch64.sil
```
These don't somehow rely on foundation at all do they?

spevans commented:
@jmittert The tests have been failing for most of the PRs for unrelated reasons

compnerd commented:
@swift-ci please test
